cou/apps/base.py

```diff
@@ -383,6 +401,8 @@ def pre_upgrade_steps(
         """
         return [
             self._get_upgrade_current_release_packages_step(units),
+            self._get_charm_hub_migration(),
+            self._get_change_to_openstack_channels(),
```

Comment:
I'm not a fan of this style: conceptually as an operator I want to prepare for an upgrade by refreshing to the latest available charm for the current OpenStack version. This can mean three things, depending on the situation (but they all fall into the same "refresh the current version" category from my point of view):
* in-channel refresh
* switch to a different channel
  * switch from charmstore to charmhub
  * switch from ch:latest/stable to a release-specific channel

The way you wrote it, you're delegating to the step-generating functions to do things if needed, which obscures the fact that the pre-upgrade steps will actually contain 2 steps, not 4.

I think it would be clearer to extend the current `_get_refresh_charm_step` function to allow switching from `latest/stable` (possibly by splitting into different nested sub-functions)

Reply:
this is better now, although I would personally go a step further and move those if conditions on the calling side. Something like
```
def _get_refresh_charm_step(self, target: OpenStackRelease) -> PreUpgradeStep:
    if self.channel == LATEST_STABLE:
        step = self._get_change_to_openstack_channels_step()
    elif not self.is_from_charm_store:
        step = self._get_charmhub_migration_step()
    # etc
    return step
```
Anyway, this is now non-blocking from my point of view